In [1]:
%matplotlib inline
%load_ext autoreload
%autoreload 2

In [2]:
from MyClassifiers import NBClassifier, LRClassifier

In [3]:
import numpy as np

In [10]:
X = np.array([[5, 6], [0.4, 1.2]])
w = np.array([0.4, 0.1])
y = np.array([0,1])

In [12]:
LRC = LRClassifier(lamda_value=1, Bias=False)
LRC.fit(X.T, y)
LRC.negative_log_likelihood(X.T, w, y)

2.3246295526739966

In [16]:
X = np.array([[1,2,3,5], [5,6,7,8]])
y = np.array([1,2,1,2])

NBC = NBClassifier()
NBC.fit(X.T, y)
print(NBC.means_)
print(NBC.var_)

[[2.  6. ]
 [3.5 7. ]]
[[1.   1.  ]
 [2.25 1.  ]]


$$ \sigma_{22} = 1 \\
    \mu_{22} = 7 \\
    \mu_{12} = 3.5  \\
    \mu_{11} = 2 \\
    \mu_{21} = 6 \\
    \sigma_{11} = 1 \\
    \sigma_{12} = 2.25 \\
    \sigma_{21} = 1 $$

In [33]:
x = np.array([1,1])
x = x[:,np.newaxis]
A = np.array([[1,3], [2,1]])

In [35]:
print((x@x.T).shape)
print(x.T @ A @ x)
print(x @ x)

(2, 2)
[[7]]


ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 1)

In [38]:
print(x.T @ x)
print(x.T @ np.linalg.inv(A) @ x)

[[2]]
[[0.6]]


In [39]:
X = np.array([[5, 6], [0.4, 1.2]])
w = np.array([0.4, 0.1])
y = np.array([0,1])

In [53]:
LCR = LRClassifier(lamda_value=1, Bias=False)
# LCR.fit(X.T, y)
LCR.gradient(X.T, w, y)

array([4.37785867, 0.36461177])

$$ m_1 = 4.37785867 \\
   m_2 = 0.36461177$$

In [49]:
X = np.array([[5, 6], [0.4, 1.2]])
w = np.array([0.4, 0.1])

In [56]:
LCR = LRClassifier(lamda_value=1, Bias=False)
LCR.hessian(X.T, w)

array([[6.02686841, 0.69989454],
       [0.69989454, 1.11554058]])

$$ H_{11} = 6.02686841 \\
   H_{12} = 0.69989454 \\
   H_{22} = 1.11554058 \\
   H_{21} = 0.69989454 $$

In [80]:
def estimate_u(x,c):
    mean_0 = 0
    n_0 = 0
    mean_1 = 0
    n_1 = 0
    for i in range(0, len(c)):
        if c[i] == 0:
            mean_0 += x[i]
            n_0 += 1
        if c[i] == 1:
            mean_1 += x[i]
            n_1 += 1
    mean_0 = mean_0 / n_0
    mean_1 = mean_1 / n_1
    
    return [mean_0, mean_1]        

In [82]:
print(["%.4f"%item for item in estimate_u([1,3,3],[0,1,0])])

['2.0000', '3.0000']
